### PR TITLE
Vitis 13170 - Allow setting thread CPU affinity for XRT applications on Windows

### PR DESCRIPTION
**Pull request by diukirde**
#### Problem solved by the commit
We want to set cpu affinity for threads on Windows like we do currently for Linux. 
#### Issue
https://jira.xilinx.com/browse/VITIS-13170
#### How problem was solved
Using the APIs provided by windows - SetThreadAffinityMask function (winbase.h)

#### What has been tested and how
cpu_affinity is being honored from xrt.ini. Verified affinity mask calculation seems to be accurate. No error in the API call observed.

cpu_affinity

#### Documentation impact (if any)
None

### CODE REVIEW

**Comment by AShivangi** [comment]
Do we want to show this in host report? We used to have it, but it has been removed since

File: src/runtime_src/core/common/thread.cpp [hunk 1]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
 }
 
 static void
-set_cpu_affinity(std::thread&)
+set_cpu_affinity(std::thread& thread)
 {
+  static bool initialized = false;
+  static DWORD_PTR affinityMask = 0;
+  static bool all = false;
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:147]: Let's fix this here and for Linux as well.  Remove this `all` variable and replace all occurrences of `all = true;` with `return;`.  Drop the outer `else` and reduce indentation with early-exit.

File: src/runtime_src/core/common/thread.cpp [hunk 2]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
...
+
+  if (!initialized) {
+    initialized = true;
+
+    std::string cpus = xrt_core::config::detail::get_string_value("Runtime.cpu_affinity", "default");
+    if (cpus == "default") {
+      all = true;
+    }
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:155]:
```suggestion
    if (cpus == "default")
      return;

```

File: src/runtime_src/core/common/thread.cpp [hunk 3]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
...
+    else {
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:156]: Remove and reduce identation.
```suggestion
```

File: src/runtime_src/core/common/thread.cpp [hunk 4]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
...
+        else {
+          xrt_core::message::send(xrt_core::message::severity_level::warning, "XRT", "Ignoring cpu affinity since cpu #" + tok + " is out of range\n");
+          all = true;
+        }
+      }
+    }
+  }
+
+  if (all)
+    return;
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:178]: Not needed.
```suggestion
```

File: src/runtime_src/core/common/thread.cpp [hunk 5]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
...
+        auto cpu = std::stoul(tok);
+        cpu = static_cast<DWORD_PTR>(cpu);
+        auto one = static_cast<DWORD_PTR>(1U);
+        if (cpu < max_cpus) {
+          XRT_DEBUG(std::cout, "adding cpu #", cpu, " to affinity mask\n");
+          affinityMask |= (one << cpu);  // Set the respective bit for the CPU
+        }
+        else {
+          xrt_core::message::send(xrt_core::message::severity_level::warning, "XRT", "Ignoring cpu affinity since cpu #" + tok + " is out of range\n");
+          all = true;
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:171]:
```suggestion
          return;
```

File: src/runtime_src/core/common/thread.cpp [hunk 6]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
 }
 
 static void
-set_cpu_affinity(std::thread&)
+set_cpu_affinity(std::thread& thread)
 {
+  static bool initialized = false;
+  static DWORD_PTR affinityMask = 0;
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:146]: user space code never uses camelCase; of course windows functions are CamelCase, but variables we create should not be.

File: src/runtime_src/core/common/thread.cpp [hunk 7]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
...
+    }
+  }
+
+  if (all)
+    return;
+
+  HANDLE threadHandle = thread.native_handle();
+  if (!SetThreadAffinityMask(threadHandle, affinityMask)) {
+    throw std::runtime_error("error calling SetThreadAffinityMask");
+  }
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:183]:
```suggestion
  if (!SetThreadAffinityMask(threadHandle, affinityMask))
    throw std::runtime_error("error calling SetThreadAffinityMask");
```

File: src/runtime_src/core/common/thread.cpp [hunk 8]
```diff
@@ -140,8 +140,47 @@ set_thread_policy(std::thread&)
...
+
+    std::string cpus = xrt_core::config::detail::get_string_value("Runtime.cpu_affinity", "default");
+    if (cpus == "default") {
+      all = true;
+    }
+    else {
+      boost::trim_if(cpus, boost::is_any_of("{}"));
+      using tokenizer = boost::tokenizer<boost::char_separator<char>>;
+      boost::char_separator<char> sep(", ");
+      auto max_cpus = GetMaximumProcessorCount(ALL_PROCESSOR_GROUPS);
```
Comment (stsoe) [inline src/runtime_src/core/common/thread.cpp:160]: Does `std::thread::hardware_concurrency` not work?

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by stsoe — APPROVED** [review]
LGTM.  Some opportunity for refactoring of code shared between Linux and Windows, but no big deal.

**Review by stsoe — APPROVED** [review]
Looks good :-)  So sorry for the confusion.

**Pull request merged**